### PR TITLE
Show specific reasons why a supply-camp cannot be placed

### PR DESCRIPTION
**Pull request by mkienenb**
2 Changes proposed in this pull request:
- Show "needs solid block", "needs air block", and "too close to existing colony" warnings along with a list of specific problem blocks (no more than 5 each) when you are unable to place a supply camp.
- Provide infrastructure for improving other placement messages (supply ship, for instance)

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 1]
```diff
@@ -912,14 +920,46 @@ private void checkAndPlace()
...
+                        placementErrorList);
+                for (final Map.Entry<PlacementErrorType, List<BlockPos>> entry : blockPosListByErrorTypeMap.entrySet()) {
+                    final PlacementErrorType placementErrorType = entry.getKey();
+                    final List<BlockPos> blockPosList = entry.getValue();
+
+                    final int numberOfBlocksTOReport = blockPosList.size() > 5 ? 5 : blockPosList.size();
+                    final List<BlockPos> blocksToReportList = blockPosList.subList(0, numberOfBlocksTOReport);
+                    String outputList = PlacementError.blockListToCommaSeparatedString(blocksToReportList);
+                    if (blockPosList.size() > numberOfBlocksTOReport) {
+                        outputList += "...";
```
[the remark this answers is not in the document]
Comment (mkienenb) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java]: Seems like a terrible idea in this case.

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 2]
```diff
@@ -912,14 +920,46 @@ private void checkAndPlace()
...
+                    }
+                    String errorMessage;
+                    switch(placementErrorType) {
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 3]
```diff
@@ -912,14 +920,46 @@ private void checkAndPlace()
...
+                    Settings.instance.getActiveStructure().getSize(BlockUtils.getRotation(Settings.instance.getRotation())), placementErrorList))
             {
                 pasteNice();
             }
             else
             {
-                LanguageHandler.sendPlayerMessage(Minecraft.getMinecraft().player, "item.supplyCampDeployer.invalid");
+                final Map<PlacementErrorType, List<BlockPos>> blockPosListByErrorTypeMap = PlacementError.partitionPlacementErrorsByErrorType(
+                        placementErrorList);
+                for (final Map.Entry<PlacementErrorType, List<BlockPos>> entry : blockPosListByErrorTypeMap.entrySet()) {
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 4]
```diff
@@ -912,14 +920,46 @@ private void checkAndPlace()
         }
         else if (FreeMode.SUPPLYCAMP == Settings.instance.getFreeMode())
         {
+            final List<PlacementError> placementErrorList = new ArrayList<PlacementError>();
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 5]
```diff
@@ -912,14 +920,46 @@ private void checkAndPlace()
...
+                final Map<PlacementErrorType, List<BlockPos>> blockPosListByErrorTypeMap = PlacementError.partitionPlacementErrorsByErrorType(
+                        placementErrorList);
+                for (final Map.Entry<PlacementErrorType, List<BlockPos>> entry : blockPosListByErrorTypeMap.entrySet()) {
+                    final PlacementErrorType placementErrorType = entry.getKey();
+                    final List<BlockPos> blockPosList = entry.getValue();
+
+                    final int numberOfBlocksTOReport = blockPosList.size() > 5 ? 5 : blockPosList.size();
+                    final List<BlockPos> blocksToReportList = blockPosList.subList(0, numberOfBlocksTOReport);
+                    String outputList = PlacementError.blockListToCommaSeparatedString(blocksToReportList);
+                    if (blockPosList.size() > numberOfBlocksTOReport) {
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    public static Map<PlacementErrorType, List<BlockPos>> partitionPlacementErrorsByErrorType(
+            List<PlacementError> placementErrorList)
+    {
+        final Map<PlacementErrorType, List<BlockPos>> blockPosListByErrorTypeMap = new HashMap<PlacementErrorType, List<BlockPos>>();
+        for (final PlacementError placementError : placementErrorList)
+        {
+            final PlacementErrorType key = placementError.getType();
+            final BlockPos blockPos = placementError.getPos();
+            List<BlockPos> blockPosList = blockPosListByErrorTypeMap.get(key);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java]: MINOR Replace this "Map.get()" and condition with a call to "Map.computeIfAbsent()". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3824)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            if (null == blockPosList)
+            {
+                blockPosList = new ArrayList<BlockPos>();
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
+package com.minecolonies.coremod.placementhandlers;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import net.minecraft.util.math.BlockPos;
+
+public class PlacementError {
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    public enum PlacementErrorType {
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            outputListStringBuilder.append(blockPos.getZ());
+            outputListStringBuilder.append(')');
+        }
+        return outputListStringBuilder.toString();
+    }
+
+    public static Map<PlacementErrorType, List<BlockPos>> partitionPlacementErrorsByErrorType(
+            List<PlacementError> placementErrorList)
+    {
+        final Map<PlacementErrorType, List<BlockPos>> blockPosListByErrorTypeMap = new HashMap<PlacementErrorType, List<BlockPos>>();
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java]: MINOR Convert this Map to an EnumMap. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1640)
MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2293)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 60 issues
* ![MAJOR][MAJOR] 50 major
* ![MINOR][MINOR] 10 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L45](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L45): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L907](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L907): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/4e0cb059607d8da18e2e3f19b8fa0701107aa25b/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.collect.Lists;
+
+import net.minecraft.util.math.BlockPos;
+
+public class PlacementError
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+{
+
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/placementhandlers/PlacementError.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 52 issues
* ![MAJOR][MAJOR] 50 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L45](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L45): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L907](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L907): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 50 issues
* ![MAJOR][MAJOR] 50 major

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L45](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L45): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L907](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L907): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/f76020c3d794392cc2a59e447e4c1b1e15451d29/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 13]
```diff
@@ -50,6 +53,11 @@
         SUPPLYCAMP
     }
 
+    private static final String SUPPLY_CAMP_INVALID_NOT_SOLID_MESSAGE_KEY = "item.supplyCampDeployer.invalid.solid_block_needed";
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java]: Move these to `TranslationConstants`.

**Review by marchermans — CHANGES_REQUESTED** [review]
Some small changes.

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 50 issues
* ![MAJOR][MAJOR] 50 major

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L903](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L903): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/28ee961d8b8069097200e7c84bc42fee729fb2f1/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Review by marchermans — APPROVED** [review]

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/2151/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Review by Raycoms — APPROVED** [review]

**Pull request merged**